---
title: "Comparison of Tissues"
author: Bryon Drown
date: January 10, 2022
output: 
  html_notebook:
    toc: true
    toc_depth: 2
---

TDMS data has been collected for four tissues (kidney, lung, heart, small intestine, and spleen).

# TODO
Bryon's todo list:

* GSEA analysis across separation tech
* Plot number of proteoforms per protein by each method
* List top proteins by number of proteoforms

```{r message=FALSE, warning=FALSE, include=FALSE}
# Load libraries

library(tidyverse)
library(VennDiagram)
library(venn)
```

A hit report was generated by TDviewer. The Hit Report includes many hits that 
correspond to PFRs that didn't pass the protein-level FDR cutoff. My solution
is to export the list of PFRs on the Proteoform tab of TDviewer and use this
list to filter the Hit Report.

```{r}
load("preprocessed.RData")
```

# QC of acquisitions

I used RawTools 2.0.3 (https://github.com/kevinkovalchik/RawTools) to analyze
the raw files.

```{r}
# Get the metrics that were generated by RawTools
metric_files <- list.files(path = "rawtools", pattern = "_Metrics.txt", full.names = TRUE)

metric_list <- lapply(metric_files, function(i){
  header <- c("RawFile", "Instrument", "ExperimentMsOrder", "MS1Analyzer", "MS2Analyzer", "MS3Analyzer", "TotalAnalysisTime", "TotalScans", 
              "NumMs1Scans", "NumMs2Scans", "NumMs3Scans", "MeanMs2TriggerRate", "Ms1ScanRate", "Ms2ScanRate", "MeanDutyCycle", "MedianMs1FillTime",
              "MedianMs2FillTime", "MedianMs3FillTime", "Ms2MedianSummedIntensity", "MedianMS1IsolationInterference", "MedianPeakWidthAt10pcH",
              "MedianPeakWidthAt50pcH", "MedianAsymmetryFactor", "PeakCapacity", "NumEsiInstabilityFlags")
  x <- read.csv(i, header = FALSE, sep = "\t")
  x <- t(x)
  colnames(x) <- header
  x[2,]
})
metrics <- bind_rows(metric_list)

metrics[,7:25] <- lapply(metrics[,7:25], function(x) as.numeric(as.character(x)))
# Parse the rawfile names
metrics <- metrics %>%
  separate(col = RawFile, into = c('Front', 'NetId', 'Project', 'Separation', 'Tissue', 'Biorep', 'Fraction', 'Injection'), 
           extra = "drop", convert = TRUE, sep = "_") %>%
  select(-Front)

metrics <- metrics %>%
  mutate(TissueSep = paste0(Tissue, " - ", Separation))
  
```


```{r}

plot <- ggplot(metrics, aes(x = MeanMs2TriggerRate,y = TissueSep, fill = TissueSep)) +
  geom_boxplot() +
  labs(x = "Mean MS2 Trigger Rate (MS2 scans per MS1 scan)") +
  theme_bw() +
  theme(legend.position = "none")
plot
pdf(file = "figures/RawTools_MS2_Trigger_Rate.pdf", width = 7, height = 5)
print(plot)
dev.off()

plot <- ggplot(metrics, aes(x = NumMs2Scans,y = TissueSep, fill = TissueSep)) +
  geom_boxplot() +
  labs(x = "Number of MS2 Scans") +
  theme_bw() +
  theme(legend.position = "none")
plot
pdf(file = "figures/RawTools_Num_MS2.pdf", width = 7, height = 5)
print(plot)
dev.off()

plot <- ggplot(metrics, aes(x = Ms2MedianSummedIntensity,y = TissueSep, fill = TissueSep)) +
  geom_boxplot() +
  labs(x = "Median MS2 Intensity") +
  theme_bw() +
  theme(legend.position = "none")
plot
pdf(file = "figures/RawTools_MS2_Intensity.pdf", width = 7, height = 5)
print(plot)
dev.off()

plot <- ggplot(metrics, aes(x = NumEsiInstabilityFlags,y = TissueSep, fill =TissueSep)) +
  geom_boxplot() +
  labs(x = "Number ESI Instability Flags") +
  theme_bw() +
  theme(legend.position = "none")
plot
pdf(file = "figures/RawTools_ESI_Instability.pdf", width = 7, height = 5)
print(plot)
dev.off()

plot <- ggplot(metrics, aes(x = MedianMs2FillTime,y = TissueSep, fill = TissueSep)) +
  geom_boxplot() +
  labs(x = "Median Ms2 Fill Time (ms)") +
  theme_bw() +
  theme(legend.position = "none")
plot
pdf(file = "figures/RawTools_MS2_Fill_Time.pdf", width = 7, height = 5)
print(plot)
dev.off()

plot <- ggplot(metrics, aes(x = MedianMs1FillTime,y = TissueSep, fill = TissueSep)) +
  geom_boxplot() +
  labs(x = "Median Ms1 Fill Time (ms)") +
  theme_bw() +
  theme(legend.position = "none")
plot
pdf(file = "figures/RawTools_MS1_Fill_Time.pdf", width = 7, height = 5)
print(plot)
dev.off()
```

# Top Proteins overall
## Proteoforms per Protein


```{r}
# First count the number of proteoforms
pfr_count <- hits %>%
  select(PFR, Uniprot_Id, Tissue) %>%
  unique() %>%
  group_by(Uniprot_Id, Tissue) %>%
  summarise(Count = n(), .groups = "drop")

# Which protein entries have the most proteoforms?
pfr_count %>%
  arrange(-Count) %>%
  pivot_wider(names_from = "Tissue", values_from = "Count")
```

There are no big surprises for which proteins are more frequently characterized


```{r}
plots <- pfr_count %>%
  group_by(Tissue) %>%
  slice_max(n = 15, Count) %>%
  do(plots = ggplot(data=.) +
       aes(x = reorder(Uniprot_Id,-Count), y = Count, label = Count) +
       geom_col() +
       geom_text(aes(y = Count + max(Count)/15)) +
       labs(title = unique(.$Tissue),
            x = "Protein Entry",
            y = "Number of Proteoforms") +
       theme_bw(base_size = 20) +
       theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1),
             plot.title = element_text(face = "bold", hjust = 0.5),
             axis.title = element_text(face = "bold"))
     ) 
plots[[2]]

pdf("figures/PFRCounts_byProtein.pdf", height = 6, width = 7)
plots[[2]]
dev.off()

```



# Unique proteoforms/proteins

## Table of unique proteoforms
For the manuscript, we need a table that describes how many proteoforms match certain criteria

```{r}
# Table 1 ======================================================================
# For the manuscript, we need a table that describes how many proteoforms match
# certain criteria

# Overall counts ###############################################################
# Proteins
df <- hits %>%
  select(HitId, PFR, Uniprot_Id, Tissue, Global_Q.value, C.score, E.value) %>%
  group_by(PFR, Tissue) %>%
  slice_min(Global_Q.value) %>% # get the best hit for each proteoform in each tissue
  slice_min(E.value) %>% # sometimes there are ties for Q.value so this breaks ties
  slice_max(C.score) %>%
  ungroup() %>%
  unique()

x <- df %>%
  select(Tissue, Uniprot_Id) %>%
  unique() %>%
  group_by(Tissue) %>%
  summarise(Protein_Total = n())
print("Number of Proteins ID overall:")
x
sum(x$Protein_Total)

# Unique proteins
proteins <- df %>%
  select(Tissue, Uniprot_Id) %>%
  unique() %>%
  group_by(Uniprot_Id) %>%
  summarise(Num_Tissues = n())

x <- df %>%
  merge(proteins) %>%
  select(Tissue, Uniprot_Id, Num_Tissues) %>%
  filter(Num_Tissues == 1) %>%
  unique() %>%
  group_by(Tissue) %>%
  summarise(Protein_Unique = n())
print("Number of Unique Proteins ID overall:")
x
sum(x$Protein_Unique)

# Proteoforms
# Total
x <- df %>%
  select(Tissue, PFR) %>%
  unique() %>%
  group_by(Tissue) %>%
  summarise(Proteoform_Total = n())
print("Number of Proteoform IDs overall:")
x
sum(x$Proteoform_Total)

# Total characterized
x <- df %>%
  select(Tissue, PFR, C.score) %>%
  filter(C.score >= 30) %>%
  unique() %>%
  group_by(Tissue) %>%
  summarise(Proteoform_Total_Char = n())
print("Number of Characterized Proteoform IDs overall:")
x
sum(x$Proteoform_Total_Char)

# Unique proteoforms
proteoforms <- df %>%
  select(Tissue, PFR) %>%
  unique() %>%
  group_by(PFR) %>%
  summarise(Num_Tissues = n())

x <- df %>%
  merge(proteoforms) %>%
  select(Tissue, PFR, Num_Tissues) %>%
  filter(Num_Tissues == 1) %>%
  unique() %>%
  group_by(Tissue) %>%
  summarise(Proteoform_Unique = n())
print("Number of Unique Proteoform IDs overall:")
x
sum(x$Proteoform_Unique)

# Unique characterized
x <- df %>%
  merge(proteoforms) %>%
  #select(Tissue, PFR, Num_Tissues, C.score) %>%
  filter(Num_Tissues == 1) %>%
  filter(C.score >= 30) %>%
  select(Tissue, PFR) %>%
  unique() %>%
  group_by(Tissue) %>%
  summarise(Proteoform_Unique_Char = n())
print("Number of Unique Characterized Proteoform IDs overall:")
x
sum(x$Proteoform_Unique_Char)

# Non-redundant counts #########################################################
# Proteins
print("Non-redundant proteins overall:")
df %>%
  select(Uniprot_Id) %>%
  unique() %>%
  nrow()

# Proteoforms
print("Non-redundant proteoforms overall:")
df %>%
  select(PFR) %>%
  unique() %>%
  nrow()

# Characterized proteoforms
print("Non-redundant characterized proteoforms overall:")
df %>%
  filter(C.score >= 30) %>%
  select(PFR) %>%
  unique() %>%
  nrow()
```

```{r}
# Tissue-specific Proteoforms --------------------------------------------------

# We want high-confidence hits
df <- hits %>%
  group_by(PFR, Tissue) %>%
  slice_min(Global_Q.value) %>% # get the best hit for each proteoform in each tissue
  slice_min(E.value) %>% # sometimes there are ties for Q.value so this breaks ties
  slice_max(C.score) %>%
  ungroup() %>%
  filter(C.score >= 30) %>%
  filter(num_fragments >= ifelse(varied | modified, 3, 0))# if the proteoform is modified or contains a sequence variance, then there should be at least 3 supporting fragments
  
# Count the number of tissues each proteoform was observed in
proteoforms <- df %>%
  select(Tissue, PFR) %>%
  unique() %>%
  group_by(PFR) %>%
  summarise(Num_Tissues = n())
  
# Select the proteoforms observed in only one tissue and prepare for export
x <- df %>%
  merge(proteoforms) %>%
  arrange(Global_Q.value) %>%
  select(Tissue, PFR, Uniprot_Id, Accession, Description, P.score, E.value, Global_Q.value, C.score, Proteoform_Level, Result_Set, Num_Tissues, varied, modified, num_fragments) %>%
  filter(Num_Tissues == 1) %>%
  select(-Num_Tissues) %>%
  unique()

# For each tissue, write out the list of uniprot ids
x %>%
  write_tsv(file = "output_data/unique_pfr_all.tsv")

x %>%
  filter(Tissue == "Heart") %>%
  write_tsv(file = "output_data/unique_pfr_heart.tsv")

x %>%
  filter(Tissue == "Lung") %>%
  write_tsv(file = "output_data/unique_pfr_lung.tsv")

x %>%
  filter(Tissue == "Spleen") %>%
  write_tsv(file = "output_data/unique_pfr_spleen.tsv")

x %>%
  filter(Tissue == "Kidney") %>%
  write_tsv(file = "output_data/unique_pfr_kidney.tsv")

x %>%
  filter(Tissue == "SmInt") %>%
  write_tsv(file = "output_data/unique_pfr_smint.tsv")

# These tsv files are used to build the xlsx file that is included in the
# supplemental materials

x %>%
  filter(Result_Set == "Tight Absolute Mass") %>%
  group_by(Tissue) %>%
  summarise(AbsoluteMass_PFRs = n())

x %>%
  filter(Result_Set == "BioMarker") %>%
  group_by(Tissue) %>%
  summarise(BioMarker_PFRs = n())

```



```{r}
# How many unique proteoforms were identified in each tissue?
x %>%
  group_by(Tissue) %>%
  summarise(PFRs = n())

x %>%
  filter(Result_Set == "BioMarker") %>%
  group_by(Tissue) %>%
  summarise(PFRs = n())

x %>%
  filter(Result_Set == "Tight Absolute Mass") %>%
  group_by(Tissue) %>%
  summarise(PFRs = n())

# How many proteins are those proteoforms associated with?
y <- x %>%
  select(Uniprot_Id, Tissue) %>%
  unique()

y %>%
  group_by(Tissue) %>%
  summarise(Proteins = n())

# Proteins associated with biomarker proteoform hits
y <- x %>%
  filter(Result_Set == "BioMarker") %>%
  select(Uniprot_Id, Tissue) %>%
  unique()

y %>%
  group_by(Tissue) %>%
  summarise(Proteins = n())

# Proteins associated with tight absolute mass hits
y <- x %>%
  filter(Result_Set == "Tight Absolute Mass") %>%
  select(Uniprot_Id, Tissue) %>%
  unique()

y %>%
  group_by(Tissue) %>%
  summarise(Proteins = n())
```

Histones and hemoglobin have the highest rate of proteoform annotation.
We also wanted to generate unique proteoform lists that exclude histones and 
hemoglobin.

```{r}
# Unique proteoforms excluding histone and hemoglobin --------------------------
# TDportal uses a curated list of histones. This is the list of accessions I
# got from Ryan Fellers.
histones <- read_csv(file = "data/Histones_Accessions_v5_BSD.txt", col_names = c("Accession"))
# While hemoglobin alpha and beta are the most frequently observed isoforms,
# I'm just including all of them.
hemoglobins <- data.frame(Accession = c("P69905", "P68871", "P69891", "P69892", "P02042"))

df <- hits %>%
  filter(!Accession %in% histones$Accession) %>%
  filter(!Accession %in% hemoglobins$Accession) %>%
  group_by(PFR, Tissue) %>%
  slice_min(Global_Q.value) %>% # get the best hit for each proteoform in each tissue
  slice_min(E.value) %>% # sometimes there are ties for Q.value so this breaks ties
  slice_max(C.score) %>%
  ungroup() %>%
  filter(C.score >= 30) %>%
  filter(num_fragments >= ifelse(varied | modified, 3, 0))# if the proteoform is modified or contains a sequence variance, then there should be at least 3 supporting fragments
  

proteoforms <- df %>%
  select(Tissue, PFR) %>%
  unique() %>%
  group_by(PFR) %>%
  summarise(Num_Tissues = n())
  
x <- df %>%
  merge(proteoforms) %>%
  arrange(Global_Q.value) %>%
  select(Tissue, PFR, Uniprot_Id, Accession, Description, P.score, E.value, Global_Q.value, C.score, Proteoform_Level, Result_Set, Num_Tissues, varied, modified, num_fragments) %>%
  filter(Num_Tissues == 1) %>%
  select(-Num_Tissues) %>%
  unique()

# For each tissue, write out the list of uniprot ids
x %>%
  write_tsv(file = "output_data/unique_pfr_all_noHH.tsv")

x %>%
  filter(Tissue == "Heart") %>%
  write_tsv(file = "output_data/unique_pfr_heart_noHH.tsv")

x %>%
  filter(Tissue == "Lung") %>%
  write_tsv(file = "output_data/unique_pfr_lung_noHH.tsv")

x %>%
  filter(Tissue == "Spleen") %>%
  write_tsv(file = "output_data/unique_pfr_spleen_noHH.tsv")

x %>%
  filter(Tissue == "Kidney") %>%
  write_tsv(file = "output_data/unique_pfr_kidney_noHH.tsv")

x %>%
  filter(Tissue == "SmInt") %>%
  write_tsv(file = "output_data/unique_pfr_smint_noHH.tsv")

# How many unique proteoforms were identified in each tissue?
x %>%
  group_by(Tissue) %>%
  summarise(PFRs = n())

x %>%
  filter(Result_Set == "BioMarker") %>%
  group_by(Tissue) %>%
  summarise(PFRs = n())

# How many proteins are those proteoforms associated with?
y <- x %>%
  select(Uniprot_Id, Tissue) %>%
  unique()

y %>%
  group_by(Tissue) %>%
  summarise(Proteins = n())

y <- x %>%
  filter(Result_Set == "BioMarker") %>%
  select(Uniprot_Id, Tissue) %>%
  unique()

y %>%
  group_by(Tissue) %>%
  summarise(Proteins = n())

```

## Comparison to Human Proteoform Atlas
Many of the proteoforms discovered in this initiative were actually 
rediscoveries judging by the PFR accession numbers assigned by TDportal.

I asked Mike to query the HPfA database and determine if the PFRs that we found
here are a part of any dataset registered in HPfA. He sent a table back with them.

```{r}
hpfa <- read.csv("data/PFRs_HPfA.txt", sep = "\t")
#ctdp <- read.csv("data/ctdp_proteoforms.csv", sep=",") %>% mutate(In.CTDP = TRUE)

df <- hits %>%
  merge(hpfa, by = "PFR", all.x = TRUE) %>%
  select(PFR, Uniprot_Id, In.HPfA) %>%
  unique() %>%
  mutate(In.HPfA = ifelse(is.na(In.HPfA), FALSE, In.HPfA)) %>%
  mutate(Dataset = ifelse(In.HPfA, "HPfA", "New"))

hpfa.dataset <-  df %>%
  filter(In.HPfA) %>%
  select(PFR, Uniprot_Id) %>% unique()
write(hpfa.dataset$PFR, "output_data/pfr_HPfA.txt", ncolumns = 1)
write(hpfa.dataset$Uniprot_Id, "output_data/protein_hpfa.txt", ncolumns = 1)

all.datasets <- df %>%
  select(PFR, Uniprot_Id) %>% unique()
write(all.datasets$PFR, "output_data/pfr_all.txt", ncolumns = 1)
write(all.datasets$Uniprot_Id, "output_data/protein_all.txt", ncolumns = 1)


plt <- venn.diagram(
  x = list(hpfa.dataset$PFR, all.datasets$PFR),
  category.names = c("HPfA", "New"),
  filename = NULL,
  
  # Title
  main = "Proteoforms in Existing Databases",
  main.cex = 1.5,
  main.fontface = "bold",
  main.fontfamily = "sans",
  
  col=c("#d95f02","#7570b3"),
  fill = c(alpha('#d95f02',0.3), alpha('#7570b3',0.3)),
  cex = 2,
  fontfamily = "sans",
  cat.cex = 2,
  cat.default.pos = "outer",
  cat.pos = c(0, 80),
  cat.dist = c(0.03, -0.03),
  cat.fontfamily = "sans",
  cat.col = c("#d95f02","#7570b3")
)
ggsave(plt, file = "figures/Dataset_PFR_Venn_all.pdf", device = "pdf")

plt <- venn.diagram(
  x = list(hpfa.dataset$Uniprot_Id, all.datasets$Uniprot_Id),
  category.names = c("HPfA", "New"),
  filename = NULL,
  
  # Title
  main = "Proteins in Existing Databases",
  main.cex = 1.5,
  main.fontface = "bold",
  main.fontfamily = "sans",
  
  col=c("#d95f02","#7570b3"),
  fill = c(alpha('#d95f02',0.3), alpha('#7570b3',0.3)),
  cex = 2,
  fontfamily = "sans",
  cat.cex = 2,
  cat.default.pos = "outer",
  cat.pos = c(0, 80),
  cat.dist = c(0.03, -0.03),
  cat.fontfamily = "sans",
  cat.col = c("#d95f02","#7570b3")
)
ggsave(plt, file = "figures/Dataset_Protein_Venn_all.pdf", device = "pdf")

# All PFRs

data <- df %>%
  select(PFR, Dataset) %>%
  unique() %>%
  group_by(Dataset) %>%
  summarise(value = n())

data <- data %>% 
  arrange(desc(Dataset)) %>%
  mutate(prop = value / sum(data$value) *100) %>%
  mutate(ypos = cumsum(prop)- 0.5*prop )
data

# Basic piechart
plot <- ggplot(data, aes(x="", y=prop, fill=Dataset)) +
  geom_bar(stat="identity", width=1, color="white") +
  coord_polar("y", start=0) +
  theme_void() + 
  scale_fill_brewer(palette="Set1")
plot
pdf(file = "figures/Dataset_PFR_Piechart_all.pdf", width = 5, height = 5)
print(plot)
dev.off()

## Proteins
data <- df %>%
  select(PFR, Uniprot_Id, In.HPfA) %>%
  unique() %>%
  group_by(Uniprot_Id) %>%
  mutate(Dataset = ifelse(TRUE %in% In.HPfA, "HPfA", "New")) %>%
  ungroup() %>%
  select(Uniprot_Id, Dataset) %>%
  unique() %>%
  group_by(Dataset) %>%
  summarise(value = n())

data <- data %>% 
  arrange(desc(Dataset)) %>%
  mutate(prop = value / sum(data$value) *100) %>%
  mutate(ypos = cumsum(prop)- 0.5*prop )
data

plot <- ggplot(data, aes(x="", y=prop, fill=Dataset)) +
  geom_bar(stat="identity", width=1, color="white") +
  coord_polar("y", start=0) +
  theme_void() + 
  scale_fill_brewer(palette="Set1")
plot
pdf(file = "figures/Dataset_Protein_Piechart_all.pdf", width = 5, height = 5)
print(plot)
dev.off()

# Histograms

df %>%
  select(PFR, In.HPfA) %>%
  unique() %>%
  ggplot(aes(x = PFR, fill = In.HPfA)) +
    geom_histogram() +
    facet_grid(~In.HPfA)
```


## Venn Diagrams

### Proteoform-level

```{r, fig.dim = c(5,5)}
venn.pfr <- hits %>%
  select(PFR, Tissue) %>%
  unique() %>%
  arrange(PFR)

heart <- filter(venn.pfr, Tissue == "Heart")$PFR
smint <- filter(venn.pfr, Tissue == "SmInt")$PFR
lung <- filter(venn.pfr, Tissue == "Lung")$PFR
kidney <- filter(venn.pfr, Tissue == "Kidney")$PFR
spleen <- filter(venn.pfr, Tissue == "Spleen")$PFR

# Save the lists to files so that external programs can access them too
write(heart, "output_data/pfr_heart.txt", ncolumns = 1)
write(smint, "output_data/pfr_smint.txt", ncolumns = 1)
write(lung, "output_data/pfr_lung.txt", ncolumns = 1)
write(kidney, "output_data/pfr_kidney.txt", ncolumns = 1)
write(spleen, "output_data/pfr_spleen.txt", ncolumns = 1)

plt <- venn.diagram(
  x = list(heart, smint, lung, kidney, spleen),
  category.names = c("Heart", "Intestine", "Lung", "Kidney", "Spleen"),
  filename = "figures/tissue_pfr_venn.tiff",
  
  # Title
  main = "Proteoforms Identified in Tissues",
  main.cex = 2,
  main.fontface = "bold",
  main.fontfamily = "sans",
  
  # Colors
  col=c('#1b9e77','#d95f02','#7570b3','#e7298a','#66a61e'),
  fill = c(alpha('#1b9e77', 0.3), alpha('#d95f02', 0.3), alpha('#7570b3', 0.3), alpha('#e7298a', 0.3), alpha('#66a61e', 0.3)),
  #fill = c(alpha("#440154ff",0.3), alpha('#21908dff',0.3)),
  cat.col=c('#1b9e77','#d95f02','#7570b3','#e7298a','#66a61e'),
  
  # Numbers
  cex = 1.3,
  fontface = "bold",
  fontfamily = "sans",
  cat.pos = c(-60,-10,-80,190,10),
  cat.dist = c(0.25, 0.25, 0.3, 0.18, 0.25),
  
  # Set names
  cat.cex = 1.5,
  cat.fontface = "bold",
  cat.default.pos = "outer",
  cat.fontfamily = "sans"
)
```

### Protein-level

```{r}
venn.protein <- hits %>%
  select(Uniprot_Id, Tissue) %>%
  unique() %>%
  arrange(Uniprot_Id)

heart <- filter(venn.protein, Tissue == "Heart")$Uniprot_Id
smint <- filter(venn.protein, Tissue == "SmInt")$Uniprot_Id
lung <- filter(venn.protein, Tissue == "Lung")$Uniprot_Id
kidney <- filter(venn.protein, Tissue == "Kidney")$Uniprot_Id
spleen <- filter(venn.protein, Tissue == "Spleen")$Uniprot_Id

# Save the lists to files so that external programs can access them too
write(heart, "output_data/protein_heart.txt", ncolumns = 1)
write(smint, "output_data/protein_smint.txt", ncolumns = 1)
write(lung, "output_data/protein_lung.txt", ncolumns = 1)
write(kidney, "output_data/protein_kidney.txt", ncolumns = 1)
write(spleen, "output_data/protein_spleen.txt", ncolumns = 1)

plt <- venn.diagram(
  x = list(heart, smint, lung, kidney, spleen),
  category.names = c("Heart", "Intestine", "Lung", "Kidney", "Spleen"),
  filename = "figures/tissue_protein_venn.tiff",
  
  # Title
  main = "Proteins Identified in Tissues",
  main.cex = 2,
  main.fontface = "bold",
  main.fontfamily = "sans",
  # Colors
  col=c('#1b9e77','#d95f02','#7570b3','#e7298a','#66a61e'),
  fill = c(alpha('#1b9e77', 0.3), alpha('#d95f02', 0.3), alpha('#7570b3', 0.3), alpha('#e7298a', 0.3), alpha('#66a61e', 0.3)),
  cat.col=c('#1b9e77','#d95f02','#7570b3','#e7298a','#66a61e'),
  
  # Numbers
  cex = 1.3,
  fontface = "bold",
  fontfamily = "sans",
  cat.pos = c(-60,-10,-80,190,10),
  cat.dist = c(0.25, 0.25, 0.3, 0.18, 0.25),
  
  # Set names
  cat.cex = 1.5,
  cat.fontface = "bold",
  cat.default.pos = "outer",
  cat.fontfamily = "sans"
)

```

## Total hits/proteins/proteoforms for each tissue

```{r}
plot <- hits %>%
  ggplot(aes(x = Tissue, fill = Tissue)) +
    geom_bar() + 
    labs(y = "Number of Hits") +
  theme_bw(base_size = 25) +
  theme(legend.position = "none")
plot
pdf(file = "figures/PfSM_byTissue.pdf")
plot
dev.off()

plot <- hits %>%
  select(Uniprot_Id, Tissue) %>%
  unique() %>%
  ggplot(aes(x = Tissue, fill = Tissue)) +
    geom_bar() + 
    labs(y = "Number of Proteins") +
  theme_bw(base_size = 25) +
  theme(legend.position = "none")
plot
pdf(file = "figures/ProteinIDs_byTissue.pdf")
plot
dev.off()

plot <- hits %>%
  select(PFR, Tissue) %>%
  unique() %>%
  ggplot(aes(x = Tissue, fill = Tissue)) +
    geom_bar() + 
    labs(y = "Number of PFRs") +
  theme_bw(base_size = 25) +
  theme(legend.position = "none")
plot
pdf(file = "figures/PFRIDs_byTissue.pdf")
plot
dev.off()

```

## Proteome Coverage from Each Tissue

```{r}
source("proteome_coverage.R")

get_coverage <- function(df) {
  df <- df %>% select(Accession, Start_Index, End_Index) %>%
    unique()
  y <- lapply(split(df[2:3], df$Accession), proteome_coverage)
  return(sum(unlist(y)))
}

df <- hits %>%
  select(Tissue, Accession, Start_Index, End_Index)
x <- lapply(split(df[2:4], df$Tissue), get_coverage)
x <- as.data.frame(x) %>%
  pivot_longer(everything(), names_to = "Tissue", values_to = "Coverage") %>%
  mutate(Percent_Coverage = Coverage/human_proteome_aa)

plot <- ggplot(x, aes(x = Tissue, y = Coverage, fill = Tissue)) +
  geom_col() +
  theme_bw(base_size = 25) +
  labs(y = "Proteome Coverage (amino acids)") +
  theme(legend.position = "none")
plot
pdf(file = "figures/ProteomCoverage_byTissue.pdf")
plot
dev.off()

plot <- ggplot(x, aes(x = Tissue, y = Percent_Coverage*100, fill = Tissue)) +
  geom_col() +
  theme_bw(base_size = 25) +
  labs(y = "Percent Proteome Coverage") +
  theme(legend.position = "none")
plot
pdf(file = "figures/PercentProteomCoverage_byTissue.pdf")
plot
dev.off()
```

```{r}
# Overall distribution of proteoform physiochemical properties

# pI
hits %>%
  select(PFR, pI, Tissue) %>%
  unique() %>%
  ggplot(aes(y = pI, x = Tissue, fill = Tissue)) +
    geom_violin() +
    theme_bw() +
    labs(title = "Proteoform pI Distribution by Tissue")

# Lipophilicity
hits %>%
  select(PFR, GRAVY, Tissue) %>%
  unique() %>%
  ggplot(aes(y = GRAVY, x = Tissue, fill = Tissue)) +
    geom_violin() +
    theme_bw() +
    labs(title = "Proteoform Lipophilicity Distribution by Tissue")

# Mass
hits %>%
  select(PFR, Monoisotopic_Mass, Tissue) %>%
  unique() %>%
  ggplot(aes(y = Monoisotopic_Mass, x = Tissue, fill = Tissue)) +
    geom_violin() +
    theme_bw() +
    labs(title = "Proteoform Mass Distribution by Tissue")



```

# PTM Frequency

```{r}
# Many of the PTMs that are actually very specific (e.g. phosphoserine)
# We instead want to group all phosphorylations together
# ptm_map relates specific PTMs to their more general PSI-MOD entries
ptm_map <- read.csv(file = "data/ptm_mapping.csv")
colnames(ptm_map)[1] <- "Modification_Codes"

df <- hits %>%
  filter(Modification_Codes == "" | num_fragments > 2 ) %>%
  select(PFR, Tissue, Modification_Codes) %>%
  unique() %>%
  separate_rows(Modification_Codes, sep="\\|") %>%
  mutate(Modification_Codes = str_split(Modification_Codes, pattern="@", simplify = TRUE)[,1])

# Now count the number of each PTM in PFRs
# If a PFR contains the same PTM multiple times, it is counted multiple times
ptm_counts <- merge(df, ptm_map)%>%
  select(General.Description, Tissue) %>%
  ungroup() %>%
  group_by(General.Description, Tissue) %>%
  summarise(Count = n(), .groups = "drop") %>%
  arrange(-Count)

# The Chi-squared test is not accurate when there are fewer than 5 observations
# so I'll remove any PTMs that were seen fewer than 5 times.
ptm_counts <- ptm_counts %>%
  pivot_wider(names_from = Tissue, values_from = Count, values_fill = 0)

# The base Chi-squared test function accepts a matrix with named rows as input
mtx <- ptm_counts %>%
  select(-General.Description) %>%
  as.matrix
rownames(mtx) <- ptm_counts$General.Description

# Let's calculate the rate of observing each mod
ptm_counts <- ptm_counts %>%
  mutate(Lung_rate = Lung / sum(ptm_counts$Lung)) %>%
  mutate(Heart_rate = Heart / sum(ptm_counts$Heart)) %>%
  mutate(Kidney_rate = Kidney / sum(ptm_counts$Kidney)) %>%
  mutate(Spleen_rate = Spleen / sum(ptm_counts$Spleen)) %>%
  mutate(SmInt_rate = SmInt / sum(ptm_counts$SmInt))

# Chi-squared
test <- chisq.test(mtx)
test

# Compare the expected distribution to the actual distribution
ptm_counts$Lung_expected <- test$expected[,"Lung"]
ptm_counts$Heart_expected <- test$expected[,"Heart"]
ptm_counts$Kidney_expected <- test$expected[,"Kidney"]
ptm_counts$Spleen_expected <- test$expected[,"Spleen"]
ptm_counts$SmInt_expected <- test$expected[,"SmInt"]
ptm_counts
```
```